src/test/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPluginTest.java

```diff
@@ -0,0 +1,191 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.conf.RuntimeProperty;
+import com.mysql.cj.jdbc.ha.ConnectionProxy;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.LRUCache;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Properties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class AWSSecretsManagerPluginTest {
+
+  private static final String SECRETS_MANAGER_REGION = "us-east-2";
+  private static final String SECRETS_MANAGER_SECRET_ID = "secretId";
+  private static final String SECRET = "{\"username\": \"testUser\", \"password\": \"testPassword\"}";
+  private static final String TEST_DB_URL = "jdbc:mysql:aws://test-domain:3306/test";
+  private static final String SQL_ERROR = "SQL exception error message";
+  private static final String SQL_STATE_FAIL_GENERIC = "HY000";
+  private AWSSecretsManagerPlugin plugin;
+  private AWSSecretsManagerPlugin spyPlugin;
+  private AutoCloseable closeable;
+
+  @Mock ConnectionProxy proxy;
+  @Mock IConnectionPlugin mockPlugin;
+  @Mock Log logger;
+  @Mock PropertySet propertySet;
+  @Mock ConnectionUrl mockConnectionUrl;
+  @Mock RuntimeProperty<String> secretsManagerRegionProperty;
+  @Mock RuntimeProperty<String> secretsManagerSecretIdProperty;
+  @Mock LRUCache<String, String> mockSecretCache;
+  @Captor ArgumentCaptor<ConnectionUrl> captor;
+
+  @BeforeEach
+  private void init() throws SQLException {
+    closeable = MockitoAnnotations.openMocks(this);
+
+    when(propertySet.getStringProperty(AWSSecretsManagerPlugin.SECRET_ID_PROPERTY))
+        .thenReturn(secretsManagerSecretIdProperty);
+    when(propertySet.getStringProperty(AWSSecretsManagerPlugin.REGION_PROPERTY))
+        .thenReturn(secretsManagerRegionProperty);
+    when(secretsManagerRegionProperty.getValue()).thenReturn(SECRETS_MANAGER_REGION);
+    when(secretsManagerSecretIdProperty.getValue()).thenReturn(SECRETS_MANAGER_SECRET_ID);
+    when(mockConnectionUrl.getDatabaseUrl()).thenReturn(TEST_DB_URL);
```

Comment:
The same concern for ConnectionUrl object. It's easy just to use it rather than mocking.